schema/representations.md

```diff
@@ -0,0 +1,256 @@
+This list representation drops the struct fieldnames from the representation
+entirely and relies purely upon the order of elements to denote meaning.
+As such, it's generally a very compact way to represent data.
+
+Tuple representations should be used cautiously.  Since they contain little
+in the way of "self-describing" information, tuple representations can make for
+very fragile protocols, increase the difficulty of migrations, and make
+serialized data incomprehensible without the schema information in hand.
+
+### struct map representation example schema with defaults
+
+```ipldsch
+type TypeList struct {
+	valueType TypeTerm
+	valueNullable Bool
+} representation map {
+	field valueNullable default "false"
+}
+```
+
+#### struct tuple representation example schema
+
+```ipldsch
+```
+
+#### struct tuple representation example data
+
+```json
+```
+
+### struct stringjoin representation example schema
+
+```ipldsch
+## Fizzlebop is a pair of fields which serializes as "part1:part2" as a string.
+type Fizzlebop struct {
+	a String
+	b String
+} representation stringjoin {
+	join ":"
+}
+```
+
+### struct stringpairs representation example schema
+
+Say we're doing something awfully like the mount options in an /etc/fstab file:
+
+```ipldsch
+## MountOptions serializes as a string of comma-separated "k=v" entries, e.g. "k1=v1,k2=v2".
+## This serial representation is limited (n.b. The domain of keys must exclude "=" and values must exclude ","!),
+## but we've chosen it because other parts of our application protocol are drastically simplified by limiting this to a string.
+type MountOptions map {String:String} representation stringpairs {
+	innerDelim "="
+	entryDelim ","
+}
+```
+
+### union keyed representation example
+
+```ipldsch
+type MyKeyedUnion union {
+	| Foo "foo"
+	| Bar "bar"
+} representation keyed
+
+type Foo struct { froz Bool }
+type Bar int
+```
+
+Some data matching `MyKeyedUnion` (shown as JSON) is:
+
+```json
+{"foo": {"froz": true}}
+```
+
+This data would also match, as the other type:
+
+```json
+{"bar": 12}
+```
+
+Note how kinded unions don't introduce any kind of wrapping object at all;
+contrast this with each of the other union representation strategies, all
+of which use at least one layer of map in their representation.
+
+### union kinded representation example
+
+```ipldsch
+type MyKindedUnion union {
+	| Foo map
+	| Bar int
+} representation kinded
+
+type Foo struct { froz Bool }
+type Bar int
+```
+
+Some data matching `MyKindedUnion` (shown as JSON) is:
+
+```json
+{"froz": true}
+```
+
+This data would also match, as the other type:
+
+```json
+12
+```
+
+Note that the syntax used in the type declaration is different for kinded
+unions versus other union representations!
+Kinded unions, unlike all other representation strategies, don't have
+any kind of string keywords at all.  Instead, they encode all of their
+discriminant information in the kind of the data itself!
```

Comment:
OK, now I get what you intended with "kinded", but it scares me because of the kind of mess that type coercion imposes on untyped languages like JavaScript. We're going to need to specify rules that differentiate the close types: int and float, if that's even possible:

```ipldsch
 type MyKindedUnion union {
	| Foo float
	| Bar int
} representation kinded
```
->
```json
1
```
or
```json
1.0
```
what do?

And how do we even handle that in JavaScript? If we decode a `1.0`, by the time it gets past the encoding layer we have some generic number type that we can't differentiate between a float and an int, it's a `1`, `typeof x === 'number'`. Is it a `MyKindedUnion#Foo` or a `MyKindedUnion#Bar`?

Reply:
we can’t make assurances about language level precision loss in this schema spec, all we can do is codify what these mean in language agnostic terms and direct each language to attempt to preserve precision as much as possible. we already do this at the data model layer in JavaScript with these types, we can’t force float types but we do try to preserve 64bit integers.

in other words, it’s fine to be as precise as possible at the schema layer as long as we understand that different languages can only make their best effort to support that precision.

Reply:
Oh, boy.  That's rough.

I'm not entirely sure.  Something like this has come up before in refmt where some care was required around int vs uint bitwidth issues, but those were... more precise and more tractable issues than this.

Reply:
I think we can get *pretty* far with some sort of rule like: the deserialization code for ambiguous cases should prefer to report the more constrained type (e.g. int over float) until forced not to by the data.  (`1.0` in JSON would be considered forcing a float.)

In javascript, it's probably going to turn out required to keep a kind tag associated with the token from the deserialization parse layer.  That's not the prettiest, but I don't see how it can be avoided.

And after all that, it still might be "not recommended" to try to use a kinded union of int-or-float because of the potential for ambiguity.

Reply:
"not recommended" sounds OK I suppose. A clear set of rules about what path is likely to be chosen when there's ambiguity might help too but a schema author is unlikely to know all possible values their code can take (or may not think this through) and might make a false assumption that all their floats will have a fractional part.
I can imagine having to let the tag bubble up from the serialization layer become a problem when we're trying to separate out these layers nicely. In fact, I'm not sure it's even possible in the JS incarnations as they exist today (old and new proposed).